### DISCUSSION TITLE
Delay on Opening Menu?

### DISCUSSION THREAD

**Discussion by harrismail** [Open-Shell/Open-Shell-Menu#1269] [General]
Hi all,
For a few Open Shell Menu "versions" I've noticed quite a delay on Windows start up when **first** clicking the Open Shell Menu start button. The delay is at least 2 seconds until the main menu opens. After that it works perfectly.
I've obviously checked the setting for Pre-Cache icons in the general settings page and that shows a tick.
It's not a biggy as, for me everything else works perfectly.
Running Windows 10 Home (x64)
Steve
![Menu](https://user-images.githubusercontent.com/34132327/210133896-36b89112-1638-4a8d-b064-c58453be9932.jpg)

**Comment by ge0rdi** [comment]
If you are able to replicate it then you can create [ETW trace](https://github.com/Open-Shell/Open-Shell-Menu/wiki/How-to-obtain-ETW-trace) and send it to me for analysis.

Ideally also enable [logging](https://github.com/Open-Shell/Open-Shell-Menu/wiki/How-to-obtain-advanced-Open-Shell-Menu-log) and provide the log too (created at the same time as the ETW trace).

**Reply by harrismail** [reply]
Hi, I've just done that but I can't seem to find the trace.etl file?

**Reply by harrismail** [reply]
Is it called trace.etl or another .etl file?

**Reply by ge0rdi** [reply]
If you followed the linked guide then the trace file should be named `trace.etl` and is created in the same folder where your console was opened in.

Did you run console as elevated administrator?
Was there some error reported?

Maybe create some new folder, switch to it (`cd folder-name`) and then do trace using `wpr` commands.

**Reply by harrismail** [reply]
Hmm, strange.
I installed the Windows ADK with default settings and ran the elevated command prompt with no errors??

**Reply by ge0rdi** [reply]
You don't need to install anything on Win10.

Just create some folder, switch to it in elevated admin console and run commands described in the ETW guide.
Eventually list outputs of those commands here.

**Reply by harrismail** [reply]
OK - I switched to another folder and that worked fine  - strange how it didn't work running an elevated cmd fromthe c:\Windows\System32\ folder.
Will get back to you.
Steve

**Reply by harrismail** [reply]
OK I've finally got round to sorting this out.
I've done an ETW trace when the start menu delay was at least 3 seconds!

Question. 
Do you want me to send just the .etl file or the folder trace.etl.NGENPDB or both?
I'll send you a WeTransfer link to ge0rdi (at) gmx.com if that's OK?

**Reply by ge0rdi** [reply]
> Do you want me to send just the .etl file

Yup, `.etl` one will be enough.

> I'll send you a WeTransfer link to ...

Perfect. Thank you.

**Reply by harrismail** [reply]
I've sent that to you.
It's really weird as sometimes there's absolutely NO delay when opening the menu and then other times it takes 2 to 5 seconds! It would be great if you could see what's going on!
Steve

**Reply by ge0rdi** [reply]
What I can see from the trace immediately is that Windows Defender is consuming cca 25% of CPU.
Not sure what exactly it is processing, but that maybe related to delays you are observing.

You may try to disable Defender or create some exclusions to see it that would help.

**Reply by harrismail** [reply]
Would that account for the “only” time the delay happens is opening the menu for the very first time after boot?
Can I exclude open shell menu from defender? Or would that not make a difference?
It’s odd because I can’t say any other thing has a pause as long as open shell menu - especially as I selected to pre-cache the menu??

**Reply by ge0rdi** [reply]
I can see cca 3.5s delay when `Open-Shell` is loading icons for start menu items. And this delay is accounted to Windows Defender (`Open-Shell` opens files and Defender scans them).

> especially as I selected to pre-cache the menu?

This is what I wanted to ask you to check, but now I see in original description that you have `Pre-cache icons` option enabled.
It should preload icons when you log-in (before the start menu is displayed).

> Would that account for the “only” time the delay happens is opening the menu for the very first time after boot?

Well, I guess so. If you open start menu right after log-in (before icons get preloaded) then Defender will have to scan them and cause delays.

There is `Delay loading of icons` setting that should cause `Open-Shell` to not wait for icons and proceed with start menu.
Do you have it enabled?

> Can I exclude open shell menu from defender? 

You can't exclude just `Open-Shell` because it is loaded (and runs) within `explorer.exe` process.
So you will need to exclude whole `explorer.exe` process from scanning. Though I don't know whether it is possible (and even good idea).

**Reply by harrismail** [reply]
Thanks for the info - the "Delay Loading of Icons" setting isn't enabled (should I try this?)

Does this kind of "delay" behaviour happen for all those using Windows Defender or is it just me ;-)

As I said - it's not a biggy as using Open Shell Menu is one thing I would definitely HATE to be without.
Thanks to all concerned for everything you guys do.

**Reply by ge0rdi** [reply]
> Thanks for the info - the "Delay Loading of Icons" setting isn't enabled (should I try this?)

You can try it. The option is supposed to be enabled by default (so it should be safe to use).

> Does this kind of "delay" behaviour happen for all those using Windows Defender or is it just me ;-)

Well, that's hard to tell, because it depends on so many factors ...
But for sure AVs may affect performance of start menu.

I have some more details about the _delay_.
In fact there is single 3.59s delay caused by defender (other Defender delays are much smaller).
During that window it seems that Defender is reading from `C:\totalcmd\TOTALCMD64.EXE` (I guess it makes sense as it seems you have it in your start menu).

So maybe try to exclude this single file from Defender's scan.

**Reply by harrismail** [reply]
That’s brilliant to find that out!! 
I’ve knocked out commander from the menu as I pretty much only use the desktop or Quick Launch icon anyway so I’ll see how it goes.
Thanks again for **all** your help 👍

**Comment by harrismail** [comment]
I've searched with Total Commander for *.etl within the past hour and these are the results.

![etl_search](https://user-images.githubusercontent.com/34132327/210136083-b6300aa3-01dd-45b7-be39-05a246c36ee8.PNG)
